ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
-    }
-
-    /// Returns `true` if the instruction defines a local variable.
-    pub(crate) fn is_def(&self) -> bool {
-        match self.opcode() {
-            OpCode::Load => true,
-            OpCode::LoadArg => true,
-        }
-    }
+/// Describes a function.
+#[derive(Clone, Debug)]
+pub(crate) struct Function {
+    /// The symbol name for the function.
+    name: String,
+    /// The type of the function.
+    tyidx: TypeIndex,
 }
 
-impl fmt::Display for Instruction {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let opc = self.opcode();
-        write!(f, "{:?}", opc)?;
-        if self.is_short() {
-            for i in 0..=2 {
-                let op = self.operand(i);
-                write!(f, "{}", op)?;
-            }
-        }
-        Ok(())
+impl Function {
+    fn new(name: String, tyidx: TypeIndex) -> Self {
+        Self { name, tyidx }
     }
 }
 
 /// The `Module` is the top-level container for JIT IR.
+///
+/// The IR is conceptually a list of word-sized instructions containing indices into auxiliary
+/// vectors.
+///
+/// The instruction stream of a [Module] is partially mutable:
+/// - you may append new instructions to the end.
+/// - you may replace and instruction with another.
+/// - you may NOT remove an instruction.
 #[derive(Debug)]
 pub(crate) struct Module {
     /// The name of the module and the eventual symbol name for the JITted code.
     name: String,
     /// The IR trace as a linear sequence of instructions.
     instrs: Vec<Instruction>,
+    /// The extra argument table.
+    ///
+    /// Used when a [CallInstruction]'s arguments don't fit inline.
+    ///
+    /// An [ExtraArgsIndex] describes an index into this.
+    extra_args: Vec<Operand>,
+    /// The type table.
+    ///
+    /// A [TypeIndex] describes an index into this.
+    types: Vec<Type>,
+    /// The function table.
+    ///
+    /// A [FunctionIndex] describes an index into this.
+    funcs: Vec<Function>,
+    /// The constant table.
+    ///
+    /// A [ConstIndex] describes an index into this.
+    consts: Vec<Constant>,
 }
 
 impl Module {
     /// Create a new [Module] with the specified name.
     pub fn new(name: String) -> Self {
-        Self {
+        let mut this = Self {
             name,
             instrs: Vec::new(),
-        }
+            extra_args: Vec::new(),
+            types: Vec::new(),
+            funcs: Vec::new(),
+            consts: Vec::new(),
+        };
+        // We reserve some type slots for common types. This means we can identify them by index
+        // without having to look into the type table.
+        //
+        // FIXME: Add more common types.
+        //
+        // The order of these pushes are important and must correspond with `TYPE_INDEX_*`;
+        this.push_type(Type::Void); // TYPE_INDEX_VOID
+        this.push_type(Type::Ptr); // TYPE_INDEX_PTR
+        this.push_type(Type::Unimplemented); // TYPE_INDEX_UNIMPLEMENTED
```

Comment:
Why not `types: vec![Type::Void, Type::Ptr, Type::Unimplemented]` rather than 3 pushes?

Reply:
No, `push_type` does a bit more than just pushing a value into a vector here.

Reply:
Actually, in this case it doesn't matter, since we are not doing anything with the index that `push_type` returns. So I guess your suggestion works.

Reply:
The pre-allocated JIT type indices were removed by the change that shares AOT type with the JIT IR: c8333b87fab7833bd6584ed4234b0d27b4c545db